mne/export/_edf.py

```diff
@@ -0,0 +1,212 @@
+    highpass = raw.info['highpass']
+    linefreq = raw.info['line_freq']
+    filter_str_info = f"HP:{highpass}Hz LP:{lowpass}Hz N:{linefreq}Hz"
+
+    # get data in uV
+    # XXX: what about non-EEG data?... How to handle?
+    units = dict()
+    units['eeg'] = 'uV'
+    units['ecog'] = 'uV'
+    units['seeg'] = 'uV'
+
+    # get the entire dataset in uV
+    data = raw.get_data(units=units, picks=ch_names)
+
+    if physical_range == 'auto':
+        # get max and min for each channel type data
+        ch_types_phys_max = dict()
+        ch_types_phys_min = dict()
+
+        for idx, ch_name in enumerate(raw.ch_names):
+            ch_type = raw.get_channel_types(picks=ch_name)[0]
+
+            ch_type_data = data[idx, :]
+            if ch_type not in ch_types_phys_max:
+                ch_types_phys_max[ch_type] = ch_type_data.max()
+                ch_types_phys_min[ch_type] = ch_type_data.min()
+            else:
+                ch_types_phys_max[ch_type] = max(
+                    ch_type_data.max(), ch_types_phys_max[ch_type])
+                ch_types_phys_min[ch_type] = min(
+                    ch_type_data.min(), ch_types_phys_min[ch_type])
+    else:
+        # get the physical min and max of the data in uV
+        # Physical ranges of the data in uV is usually set by the manufacturer
+        # and properties of the electrode. In general, physical max and min
+        # should be the clipping levels of the ADC input and they should be
+        # the same for all channels. For example, Nihon Kohden uses +3200 uV
+        # and -3200 uV for all EEG channels (which are the actual clipping
+        # levels of their input amplifiers & ADC).
+        # For full discussion, see: https://github.com/sccn/eeglab/issues/246
+        pmin, pmax = -3200, 3200
+
+        # check that physical min and max is not exceeded
+        if data.max() > pmax:
+            raise RuntimeError(f'The maximum uV of the data {data.max()} '
+                               f'is more then physical max passed in {pmax}.')
+        if data.min() < pmin:
+            raise RuntimeError('The minimum uV of the data {data.min()} '
+                               f'is less then physical min passed in {pmin}.')
+
+    # create instance of EDF Writer
+    hdl = EDFwriter(fname, EDFwriter.EDFLIB_FILETYPE_EDFPLUS, n_chs)
+
+    # set channel data
+    for ichan, ch in enumerate(ch_names):
+        if physical_range == 'auto':
+            # take the channel type minimum and maximum
+            ch_type = raw.get_channel_types(picks=ch)[0]
+            pmin, pmax = ch_types_phys_min[ch_type], ch_types_phys_max[ch_type]
+
+        for key, val in zip(
+            ['PhysicalMaximum', 'PhysicalMinimum', 'DigitalMaximum',
+             'DigitalMinimum', 'PhysicalDimension', 'SampleFrequency',
+             'SignalLabel', 'PreFilter',
+             ],
+            [pmax, pmin, digital_max, digital_min, phys_dims, sfreq,
+             ch, filter_str_info]
```

Comment:
I would find looping over a list of tuples of len 2 much less error prone. Here if you swap entries in a list you have a silent bug which is hard to see